### PR TITLE
planner: adjust estimated row count when pushing `Limit` and keep order for partitioned table

### LINKED ISSUE #40986: Adjust the estimated row count when push `Limit` and keep order for partitioned table
## Enhancement

When there's a `Limit` above a scan operator, we should adjust the estimated row count of the scan operator accordingly.

But when we push `Limit` and keep order for a partitioned table (introduced in https://github.com/pingcap/tidb/pull/36108 as a temporary enhancement), we didn't do this:

```sql
use test;
create table t (a int, b int, c int, index ia(a)) partition by range(b) (partition p1 values less than(100), partition p2 values less than(200), partition p3 values less than maxvalue);
analyze table t;
explain select a from t where a > 10 order by a limit 10;
```
```
+-----------------------------+---------+-----------+----------------------+-------------------------------------------------+
| id                          | estRows | task      | access object        | operator info                                   |
+-----------------------------+---------+-----------+----------------------+-------------------------------------------------+
| TopN_8                      | 10.00   | root      |                      | test.t.a, offset:0, count:10                    |
| └─IndexReader_15            | 10.00   | root      | partition:all        | index:Limit_14                                  |
|   └─Limit_14                | 10.00   | cop[tikv] |                      | offset:0, count:10                              |
|     └─IndexRangeScan_13     | 3333.33 | cop[tikv] | table:t, index:ia(a) | range:(10,+inf], keep order:false, stats:pseudo |
+-----------------------------+---------+-----------+----------------------+-------------------------------------------------+
```

Normally, if it's not a partitioned table, the estimated row count is like this:
```sql
use test;
create table t1 (a int, b int, c int, index ia(a));
analyze table t1;
explain select a from t1 order by a limit 10;
```
```
+----------------------------+---------+-----------+-----------------------+-------------------------------+
| id                         | estRows | task      | access object         | operator info                 |
+----------------------------+---------+-----------+-----------------------+-------------------------------+
| Limit_10                   | 10.00   | root      |                       | offset:0, count:10            |
| └─IndexReader_18           | 10.00   | root      |                       | index:Limit_17                |
|   └─Limit_17               | 10.00   | cop[tikv] |                       | offset:0, count:10            |
|     └─IndexFullScan_16     | 10.00   | cop[tikv] | table:t1, index:ia(a) | keep order:true, stats:pseudo |
+----------------------------+---------+-----------+-----------------------+-------------------------------+

```

### PR DESCRIPTION
**Pull request by time-and-fate**
### What problem does this PR solve?

Issue Number: close #40986

Problem Summary:

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: planner/core/task.go [hunk 1]
```diff
@@ -1083,6 +1097,17 @@ func (p *PhysicalTopN) pushTopNDownToDynamicPartition(copTsk *copTask) (task, bo
 		}.Init(p.SCtx(), stats, p.SelectBlockOffset())
 		pushedLimit.SetSchema(copTsk.indexPlan.Schema())
 		copTsk = attachPlan2Task(pushedLimit, copTsk).(*copTask)
+
+		// A similar but simplified logic compared the ExpectedCnt handling logic in getOriginalPhysicalIndexScan.
+		child := pushedLimit.Children()[0]
+		// The row count of the direct child of Limit should be adjusted to be no larger than the Limit.Count.
+		child.SetStats(child.statsInfo().ScaleByExpectCnt(float64(newCount)))
```
Comment (winoros) [inline planner/core/task.go:1104]: We can call `ScaleByExpectCnt` directly? That func changed the pointer itself.

File: planner/core/task.go [hunk 2]
```diff
@@ -1056,12 +1061,21 @@ func (p *PhysicalTopN) pushTopNDownToDynamicPartition(copTsk *copTask) (task, bo
...
 		finalTblScanPlan := copTsk.tablePlan
 		for len(finalTblScanPlan.Children()) > 0 {
+			selOnTblScan, _ = finalTblScanPlan.(*PhysicalSelection)
 			finalTblScanPlan = finalTblScanPlan.Children()[0]
 		}
 		tblScan = finalTblScanPlan.(*PhysicalTableScan)
 		tblInfo = tblScan.Table
 	}
 
+	// Note that we only need to care about one Selection at most.
```
Comment (AilinKid) [inline planner/core/task.go:1071]: only care about specified tree type?

**Comment by AilinKid** [comment]
rest LGTM

File: planner/core/task.go (hunk 1 above)
Comment (time-and-fate) [reply planner/core/task.go:1104]: > We can call `ScaleByExpectCnt` directly? That func changed the pointer itself.

No. It won't change itself.

File: planner/core/task.go (hunk 2 above)
Comment (time-and-fate) [reply planner/core/task.go:1071]: Yes. From the current implementation, the only thing that could possibly be here is a `Selection`.
@winoros can you verify this?

**Review by AilinKid — APPROVED** [review]

File: planner/core/task.go (hunk 2 above)
Comment (winoros) [reply planner/core/task.go:1071]: yes

**Review by winoros — APPROVED** [review]

**Comment by hawkingrei** [comment]
/hold 

CI is too busy.

**Pull request merged**